File: app/src/main/java/com/jsqix/dongqing/gank/fragment/RecommendFragment.java
package com.jsqix.dongqing.gank.fragment;

import android.support.v7.widget.LinearLayoutManager;
import android.util.Log;

import com.google.gson.Gson;
import com.jsqix.dongqing.gank.R;
import com.jsqix.dongqing.gank.adapter.recycler.RecyclerRecAdapter;
import com.jsqix.dongqing.gank.api.Api;
import com.jsqix.dongqing.gank.bean.GankData;
import com.jsqix.dongqing.gank.bean.ImgDataBean;
import com.jsqix.utils.ACache;
import com.lcodecore.tkrefreshlayout.RefreshListenerAdapter;
import com.lcodecore.tkrefreshlayout.TwinklingRefreshLayout;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Func1;
import rx.schedulers.Schedulers;

/**
 * Created by dongqing on 2016/12/8.
 * 每日数据
 */

public class RecommendFragment extends BaseFragment {
    private RecyclerRecAdapter adapter;
    private List<GankData.ResultsBean> data = new ArrayList<>();

    final static long DAY_OF_MILLISECOND = 24 * 60 * 60 * 1000L;
    private Date mCurrentDate;

    public RecommendFragment() {
        // Required empty public constructor
    }


    protected void initView() {
        LinearLayoutManager manager = new LinearLayoutManager(mContext, LinearLayoutManager.VERTICAL, false);
        recyclerView.setLayoutManager(manager);
        adapter = new RecyclerRecAdapter(mContext, R.layout.item_recycler_rec, data);
        recyclerView.setAdapter(adapter);

        refreshLayout.setEnableLoadmore(false);
        refreshLayout.setOverScrollBottomShow(false);
        refreshLayout.setOnRefreshListener(new RefreshListenerAdapter() {
            @Override
            public void onRefresh(TwinklingRefreshLayout refreshLayout) {
                page = 1;
                getData(new Date());
            }

            @Override
            public void onLoadMore(TwinklingRefreshLayout refreshLayout) {
                page++;
                getData(mCurrentDate);
            }
        });
    }

    private void getData(final Date date) {
        Calendar calender = Calendar.getInstance();
        calender.setTime(date);
        Api.getInstance().service.getDayData(calender.get(Calendar.YEAR), calender.get(Calendar.MONTH) + 1, calender.get(Calendar.DAY_OF_MONTH))
                .subscribeOn(Schedulers.io())
                .map(new Func1<GankData, GankData.ResultsBean>() {
                    @Override
                    public GankData.ResultsBean call(GankData gankData) {
                        return gankData.getResults();
                    }
                })
//                .map(new Func1<GankData.ResultsBean, List<ImgDataBean>>() {
//
//                    @Override
//                    public List<ImgDataBean> call(GankData.ResultsBean resultsBean) {
//                        return addAllResults(resultsBean);
//                    }
//                })
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Subscriber<GankData.ResultsBean>() {

                    @Override
                    public void onCompleted() {
                        Log.v("success:", "onCompleted");
                        mCurrentDate = new Date(date.getTime() - DAY_OF_MILLISECOND);
                    }

                    @Override
                    public void onError(Throwable e) {
                        Log.v("failure:", "onError");
                        refreshLayout.finishRefreshing();
                        refreshLayout.finishLoadmore();
                    }

                    @Override
                    public void onNext(GankData.ResultsBean resultsBean) {
                        if ("{}".equals(new Gson().toJson(resultsBean))) {
                            getData(new Date(date.getTime() - DAY_OF_MILLISECOND));
                        } else {
                            if (page == 1) {
                                data.clear();
                                refreshLayout.finishRefreshing();
                            } else {
                                refreshLayout.finishLoadmore();
                            }

                            data.add(resultsBean);
                            if (page == 1) {
                                adapter.refresh();
                                if (resultsBean.get福利() != null && resultsBean.get福利().size() > 0)
                                    ACache.get(mContext).put("imgUrl", resultsBean.get福利().get(0).getUrl());
                            } else {
                                adapter.notifyDataSetChanged();
                            }
                        }
                    }

                });
    }

    private List<ImgDataBean> addAllResults(GankData.ResultsBean resultsBean) {
        return null;
    }
}

File: app/src/main/java/com/dongqing/gank/kotlin/ui/fragment/RecommendFragment.kt
package com.dongqing.gank.kotlin.ui.fragment

import android.support.v7.widget.LinearLayoutManager
import android.text.TextUtils
import android.util.Log
import com.dongqing.gank.kotlin.bean.gank.FuliBean
import com.dongqing.gank.kotlin.bean.gank.GankData
import com.dongqing.gank.kotlin.bean.gank.HeaderBean
import com.dongqing.gank.kotlin.bean.gank.RecentBean
import com.dongqing.gank.kotlin.ui.adapter.multi.FuliViewBinder
import com.dongqing.gank.kotlin.ui.adapter.multi.RecentViewBinder
import com.dongqing.gank.kotlin.ui.present.RecommendPresent
import com.dongqing.gank.kotlin.ui.view.IRecommendView
import com.dongqing.gank.kotlin.utils.ACache
import com.google.gson.Gson
import kotlinx.android.synthetic.main.fragment_base.*
import me.drakeet.multitype.Items
import me.drakeet.multitype.MultiTypeAdapter
import java.util.*


class RecommendFragment : RefreshFragment<RecommendPresent, IRecommendView>(), IRecommendView {
    companion object {
        const val DAY_OF_MILLISECOND = 24 * 60 * 60 * 1000L
    }

    private var mCurrentDate = Date()
    lateinit var items: Items
    lateinit var adapter: MultiTypeAdapter
    override fun initView() {
        super.initView()
        recyclerView.layoutManager = LinearLayoutManager(mContext, LinearLayoutManager.VERTICAL, false)
        items = Items()
        adapter = MultiTypeAdapter()
        adapter.register(FuliBean::class.java, FuliViewBinder())
        adapter.register(RecentBean::class.java, RecentViewBinder())
        recyclerView.adapter = adapter
    }

    override fun getData() {
        if (refrsh) {
            mCurrentDate = Date()
        }
        val calender = Calendar.getInstance()
        calender.time = mCurrentDate
        mPresenter!!.getDayData(calender[Calendar.YEAR], calender[Calendar.MONTH] + 1, calender[Calendar.DAY_OF_MONTH])
    }

    override fun onDataReceive(t: GankData.ResultsBean) {
        refrsh = false
        var result = Gson().toJson(t)
        Log.d("day gank", result)
        if (TextUtils.equals("{}", result)) {
            mCurrentDate = Date(mCurrentDate.time - DAY_OF_MILLISECOND)
            getData()
        } else {
            if (page == 1) {
                items.clear()
                if (refresh_layout != null) {
                    refresh_layout.finishRefreshing()
                }
            }
            if (refresh_layout != null) {
                refresh_layout.finishLoadmore()
            }
            if (t != null) {
                if (t.福利 != null && t.福利.size > 0) {
                    val img = t.福利[0]
                    var gson = Gson()
                    val bean = gson.fromJson(gson.toJson(img), FuliBean::class.java)
                    items.add(bean)
                    ACache.get(mContext).put("imgUrl", img.url)
                }
                if (t.android != null && t.android.size > 0) {
                    var bean = RecentBean()
                    bean.headerBean = HeaderBean("Android")
                    bean.list = t.android
                    items.add(bean)
                }
                if (t.getiOS() != null && t.getiOS().size > 0) {
                    var bean = RecentBean()
                    bean.headerBean = HeaderBean("iOS")
                    bean.list = t.getiOS()
                    items.add(bean)
                }
                if (t.前端 != null && t.前端.size > 0) {
                    var bean = RecentBean()
                    bean.headerBean = HeaderBean("前端")
                    bean.list = t.前端
                    items.add(bean)
                }
                if (t.瞎推荐 != null && t.瞎推荐.size > 0) {
                    var bean = RecentBean()
                    bean.headerBean = HeaderBean("瞎推荐")
                    bean.list = t.瞎推荐
                    items.add(bean)
                }
                if (t.拓展资源 != null && t.拓展资源.size > 0) {
                    var bean = RecentBean()
                    bean.headerBean = HeaderBean("拓展资源")
                    bean.list = t.拓展资源
                    items.add(bean)
                }
                if (t.休息视频 != null && t.休息视频.size > 0) {
                    var bean = RecentBean()
                    bean.headerBean = HeaderBean("休息视频")
                    bean.list = t.休息视频
                    items.add(bean)
                }
                adapter.items = items
                adapter.notifyDataSetChanged()
                if (page == 1) {
                    recyclerView.smoothScrollToPosition(0)
                }
            }
        }
    }

    override fun onDataError() {
        if (refresh_layout != null) {
            refresh_layout.finishRefreshing()
            refresh_layout.finishLoadmore()
        }
    }

    override fun onUpdateDate() {
        mCurrentDate = Date(mCurrentDate.time - DAY_OF_MILLISECOND)
    }
}